### DISCUSSION TITLE
Optional route parameters are not yet supported. Is it a bug?

### DISCUSSION THREAD

**Discussion by emrecoban** [vercel/next.js#50272] [Help]
### Summary

I am creating dynamic segments with Next.js. However, if the segment is `[[slug]]`, I am getting the error "Optional route parameters are not yet supported". When it's `[[...slug]]`, I don't get the error. It's very strange. What is the solution?

My directory structure: `src/app/[[slug]]`

**Comment by icyJoseph** [comment]
What that ever supported? https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes

I do think, catch all have to be `[[...slug]]`, as it says there in the docs:

> For example, pages/shop/[[...slug]].js will also match /shop, in addition to /shop/clothes, /shop/clothes/tops, /shop/clothes/tops/t-shirts.

**Reply by emrecoban** [reply]
Yes, I saw it, but I don't want to catch all-segments. When it is, it is possible `domain.com/a/b/c/d...` I just want to create one segment, like this: `domain.com/a`

**Reply by icyJoseph** [reply]
But you can just check if the length of the captured slug is one, otherwise return not found

**Reply by emrecoban** [reply]
Yes, I know and did it like that. But It was not what I want. It should have worked like `/app/[[slug]]`.

**Reply by icyJoseph** [reply]
Ah I kind of agree. A couple of times, I've been bitten by that haha; I think the old path resolver implementation, could also handle it, but for some reason it was not allowed anyway.

**Comment by milovangudelj** [comment]
Another thing to note is that a catch-all segment must be the last part of the URL. While having support for an optional route parameter like `[[param]]` could be very useful especially for i18n where you would want the default locale of a page to just be something like this `site.com/about` instead of `site.com/en/about`.

**Reply by bahag-stolyarchukm** [reply]
Exactly what I am looking at the moment as well. Perhaps @milovangudelj you found something temporary or permanent for the case you have mentioned?

**Reply by milovangudelj** [reply]
No, currently there is nothing other than a rewrite that allows us to mask the default locale in the url.

You have to use a dynamic segment and generate all possible parameters using `generateStaticParams`. Then for each request verify if the user is visiting the default locale and rewrite the request accordingly.

There's a pretty good official example of a barebones i18n implementation here:

https://github.com/vercel/next.js/tree/canary/examples/app-dir-i18n-routing

I'll update it to add the locale masking logic as soon as I get the chance to do so.

**Reply by bahag-stolyarchukm** [reply]
Rewrite you mean by using `NextResponse.rewrite()` ?

**Reply by mfkrause** [reply]
@bahag-stolyarchukm Did you solve this in the meantime? This is exactly my problem as well right now (default locale at the base path with sub-path routing to other locales).

**Reply by bahag-stolyarchukm** [reply]
I am not sure if I went exactly the same way as @milovangudelj - but I used route masking approach. I have also used redirect approach as sometimes, based on the need, URL contains `/locale` 

```jsx
const absoluteUrl = new URL('[any route to mask]', req.nextUrl.origin).toString()

    const i18cookie = { name: cookieName, value: 'de' }

    const response = NextResponse.rewrite(absoluteUrl)

    response.cookies.set(i18cookie)

    return response 
```

What I did not use is `generateStaticParams` - as we are detecting supported `locale` on `i18n` side with `Navigator` and passing the value directly into configuration. In this way `useTranslation` does not need 2 arguments.

**Reply by bahag-stolyarchukm** [reply]
I have actually decided to go a different way.

next.config.ts:
```jsx
  async rewrites() {
    if (!isLocale) {
      return [
        {
          source: '/:path*',
          destination: '/your locale OR any segment locale like :loca/:path*'
        }
      ]
    }
  }
```

I could still be wrong, however, it works as expected for the time being.

**Comment by treylon** [comment]
Is there any news whether this is on the roadmap or not? As @milovangudelj mentioned we would also really appreciate this feature for i18n purposes.

**Comment by bahag-stolyarchukm** [comment]
It is not yet supported, is not it?

**Comment by musjj** [comment]
Can't believe that this isn't supported yet. That darn message was added 4 YEARS ago: [`f8125c9`](https://github.com/vercel/next.js/commit/f8125c9ed7db6cf78385cd62bd46ebd4de88ea49#diff-43ee6b505c851f4532c22a6082aa90fadccb347d0ad4bc93d070b9b78ef367fcR168). Now that we're on v15, still nothing.

**Comment by HackerQED** [comment]
+1. This feature is critical for i18n sites

**Comment by provokcx** [comment]
@rauchg 👀

**Reply by milovangudelj** [reply]
Please don't tag people not involved in the discussion. It's considered rude...

**Reply by provokcx** [reply]
Okay, I understand @milovangudelj
I won't do it again 👌 

So, since you're the acting marshal of this town, how do you plan on getting an answer to this 4-year-old problem?

**Reply by milovangudelj** [reply]
Don't take it personally, I'm no marshall. I just told you it's something to avoid most of the time and you can do what you want with that information 🙌🏻

As for getting the issue answered, I can only hope enough people see it and react to prioritise it. At the end of the day it's up to the maintainers to decide what to do. This is not a simple fix, and it's not even a bug really so it might never be addressed.

It would be nice but we can only wait and hope they see it 🤞🏻

**Reply by provokcx** [reply]
Yeah, no problem @milovangudelj. It's just my way of talking. I'm blunt 😂 

Let's hope more people react to it then.

**Comment by akarabach** [comment]
any updates on it ?

**Reply by dominik-sfl** [reply]
also curious about this? any news?

**Comment by sharmapukar217** [comment]
is there any new news about this?
i really need app/dashboard/[[tenant]] for this. sveltekit so nicely support this feature, i.e: routes/dashboard/[[slug]]. I've been waiting since nextjs 14, but i don't see any new topic related to this.

**Comment by AzizStark** [comment]
Aug 21

**Comment by raRaRa** [comment]
This would be a really nice feature for NextJS 16 😅

**Comment by picardplaisimond** [comment]
2026, and this is still a problem...

**Reply by sharmapukar217** [reply]
I guess it's never coming 😂
